From ac04f0950388b3e158d3c14597545d4ba8f6ed46 Mon Sep 17 00:00:00 2001
From: Priyanka Bose <priyanka.bose@intel.com>
Date: Sun, 25 Aug 2019 10:44:02 +0530
Subject: [PATCH] Adding AVX2 Implementation of Math Func:exp()

This implementation is for x86_64 platforms having the support of AVX2

Change-Id: I9d5bc41dd59dd88d36ca6ba77c082efa16703ae6
Signed-off-by: Priyanka Bose <priyanka.bose@intel.com>
---
 libm/Android.bp              |    5 +-
 libm/x86_64/exp_avx2.S       |  152 +++++
 libm/x86_64/exp_cout_avx2.S  |  250 ++++----
 libm/x86_64/exp_data_avx2.S  | 1408 +++++++++++++++---------------------------
 libm/x86_64/expf_cout_avx2.S |  548 ++++++++++++++++
 libm/x86_64/expf_data_avx2.S |  958 ++++++++++++++++++++++++++++
 6 files changed, 2287 insertions(+), 1034 deletions(-)
 create mode 100644 libm/x86_64/exp_avx2.S
 create mode 100644 libm/x86_64/expf_cout_avx2.S
 create mode 100644 libm/x86_64/expf_data_avx2.S

diff --git a/libm/Android.bp b/libm/Android.bp
index 84ec670..db5b628 100644
--- a/libm/Android.bp
+++ b/libm/Android.bp
@@ -474,7 +474,10 @@ cc_library {
                     "x86_64/rintf_avx2.S",
                     "x86_64/exp_data_avx2.S",
                     "x86_64/exp_cout_avx2.S",
-                    "x86_64/expf_avx2.S",
+                    "x86_64/exp_avx2.S",
+                    "x86_64/expf_data_avx2.S",
+                    "x86_64/expf_cout_avx2.S",
+                    "x86_64/expf_avx2.S",           
 
                     "x86_64/hypot_data_avx2.S",
                     "x86_64/hypot_cout_avx2.S",
diff --git a/libm/x86_64/exp_avx2.S b/libm/x86_64/exp_avx2.S
new file mode 100644
index 0000000..af35265
--- /dev/null
+++ b/libm/x86_64/exp_avx2.S
@@ -0,0 +1,152 @@
+/*
+Copyright (C) 2019 The Android Open Source Project
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions
+are met:
+ * Redistributions of source code must retain the above copyright
+   notice, this list of conditions and the following disclaimer.
+ * Redistributions in binary form must reproduce the above copyright
+   notice, this list of conditions and the following disclaimer in
+   the documentation and/or other materials provided with the
+   distribution.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT
+OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+SUCH DAMAGE.
+*/
+
+#include <private/bionic_asm.h>
+# -- Begin  exp
+ENTRY(exp)
+# parameter 1: %xmm0
+..B1.1:
+..L2:
+
+        pushq     %rbp
+	.cfi_def_cfa_offset 16
+        movq      %rsp, %rbp
+	.cfi_def_cfa 6, 16
+	.cfi_offset 6, -16
+        andq      $-64, %rsp
+        pushq     %rsi
+        pushq     %rdi
+        subq      $176, %rsp
+        vmovapd   %xmm0, %xmm1
+        vmovsd    1088+__avx2_exp_data_internal(%rip), %xmm2
+        lea       __avx2_exp_data_internal(%rip), %rdx
+        vfmadd132sd 1024+__avx2_exp_data_internal(%rip), %xmm2, %xmm1
+        vmovsd    1728+__avx2_exp_data_internal(%rip), %xmm3
+        vandpd    %xmm3, %xmm0, %xmm4
+        vmovups   %xmm9, 16(%rsp)
+        vmovups   %xmm8, 32(%rsp)
+        vmovups   %xmm10, 128(%rsp)
+        vsubsd    %xmm2, %xmm1, %xmm1
+        vcmpnlesd 1792+__avx2_exp_data_internal(%rip), %xmm4, %xmm5
+        vmovapd   %xmm1, %xmm7
+        vfnmadd132sd 1216+__avx2_exp_data_internal(%rip), %xmm0, %xmm7
+        vmovsd    1536+__avx2_exp_data_internal(%rip), %xmm2
+        vmovq     1664+__avx2_exp_data_internal(%rip), %xmm4
+        vmovmskpd %xmm5, %ecx
+        vaddsd    1152+__avx2_exp_data_internal(%rip), %xmm1, %xmm5
+        vfnmadd132sd 1280+__avx2_exp_data_internal(%rip), %xmm7, %xmm1
+        vpand     %xmm4, %xmm5, %xmm6
+        vpandn    %xmm5, %xmm4, %xmm8
+        vpsllq    $42, %xmm8, %xmm10
+        vfmadd213sd 1472+__avx2_exp_data_internal(%rip), %xmm1, %xmm2
+        vmulsd    %xmm1, %xmm1, %xmm7
+        vmovd     %xmm6, %eax
+        vfmadd213sd 1408+__avx2_exp_data_internal(%rip), %xmm1, %xmm2
+        vmovups   %xmm15, 96(%rsp)
+        vmovups   %xmm14, 144(%rsp)
+        vmovups   %xmm13, 48(%rsp)
+        vmovups   %xmm12, 80(%rsp)
+        vfmadd213sd 1344+__avx2_exp_data_internal(%rip), %xmm1, %xmm2
+        vmovups   %xmm11, 112(%rsp)
+        movslq    %eax, %rax
+        vfmadd213sd %xmm1, %xmm7, %xmm2
+        vmovupd   (%rdx,%rax), %xmm3
+        vmovddup  8(%rdx,%rax), %xmm6
+        vfmadd213sd %xmm6, %xmm3, %xmm2
+        vaddsd    %xmm3, %xmm2, %xmm9
+        vpaddq    %xmm10, %xmm9, %xmm1
+        testb     $1, %cl
+        jne       ..B1.3
+..B1.2:
+        vmovups   32(%rsp), %xmm8
+	.cfi_restore 25
+        vmovapd   %xmm1, %xmm0
+        vmovups   16(%rsp), %xmm9
+	.cfi_restore 26
+        vmovups   128(%rsp), %xmm10
+	.cfi_restore 27
+        vmovups   112(%rsp), %xmm11
+	.cfi_restore 28
+        vmovups   80(%rsp), %xmm12
+	.cfi_restore 29
+        vmovups   48(%rsp), %xmm13
+	.cfi_restore 30
+        vmovups   144(%rsp), %xmm14
+	.cfi_restore 31
+        vmovups   96(%rsp), %xmm15
+	.cfi_restore 32
+        addq      $176, %rsp
+	.cfi_restore 5
+        popq      %rdi
+	.cfi_restore 4
+        popq      %rsi
+        movq      %rbp, %rsp
+        popq      %rbp
+	.cfi_def_cfa 7, 8
+	.cfi_restore 6
+        ret
+	.cfi_def_cfa 6, 16
+	.cfi_offset 6, -16
+..B1.3:
+        vcmpgtsd  1984+__avx2_exp_data_internal(%rip), %xmm0, %xmm2
+        vcmpltsd  2048+__avx2_exp_data_internal(%rip), %xmm0, %xmm3
+        vblendvpd %xmm2, .L_2il0floatpacket.15(%rip), %xmm1, %xmm1
+        vorpd     %xmm3, %xmm2, %xmm4
+        vmovmskpd %xmm4, %eax
+        vandnpd   %xmm1, %xmm3, %xmm1
+        andn      %ecx, %eax, %eax
+        andl      $1, %eax
+        je        ..B1.2
+..B1.4:
+        vmovsd    %xmm0, (%rsp)
+        vmovsd    %xmm1, 64(%rsp)
+        jne       ..B1.7
+..B1.5:
+        vmovsd    64(%rsp), %xmm1
+        jmp       ..B1.2
+..B1.7:
+        lea       (%rsp), %rdi
+        lea       64(%rsp), %rsi
+..___tag_value___svml_exp1_ha_l9.55:
+        call      __avx2_exp_cout_rare_internal
+..___tag_value___svml_exp1_ha_l9.56:
+        jmp       ..B1.5
+END(exp)
+# -- End exp
+	.section .rodata, "a"
+	.align 16
+	.align 16
+.L_2il0floatpacket.15:
+	.long	0x00000000,0x7ff00000,0x00000000,0x7ff00000
+	.type	.L_2il0floatpacket.15,@object
+	.size	.L_2il0floatpacket.15,16
+	.data
+	.hidden __avx2_exp_data_internal
+	.hidden __avx2_exp_cout_rare_internal
+	.section .note.GNU-stack, ""
+# End
diff --git a/libm/x86_64/exp_cout_avx2.S b/libm/x86_64/exp_cout_avx2.S
index 2fd42c3..0b5d2ed 100644
--- a/libm/x86_64/exp_cout_avx2.S
+++ b/libm/x86_64/exp_cout_avx2.S
@@ -26,147 +26,136 @@ OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 SUCH DAMAGE.
 */
 
-# -- Machine type EM64t
-	.text
-..TXTST0:
-.L_2__routine_start___svml_sexp_ha_cout_rare_internal_0:
-# -- Begin  __svml_sexp_ha_cout_rare_internal
-	.text
-       .align    16,0x90
-	.hidden __svml_sexp_ha_cout_rare_internal
-	.globl __svml_sexp_ha_cout_rare_internal
-__svml_sexp_ha_cout_rare_internal:
+#include <private/bionic_asm.h>
+# -- Begin avx2_exp_cout_rare_internal
+ENTRY(__avx2_exp_cout_rare_internal)
 # parameter 1: %rdi
 # parameter 2: %rsi
 ..B1.1:
-	.cfi_startproc
-..___tag_value___svml_sexp_ha_cout_rare_internal.1:
-..L2:
+..L1:
 
         xorl      %eax, %eax
-        movzwl    2(%rdi), %edx
-        andl      $32640, %edx
-        shrl      $7, %edx
-        cmpl      $255, %edx
+        movzwl    6(%rdi), %edx
+        andl      $32752, %edx
+        shrl      $4, %edx
+        movsd     (%rdi), %xmm6
+        movsd     %xmm6, -8(%rsp)
+        cmpl      $2047, %edx
         je        ..B1.17
 ..B1.2:
-        pxor      %xmm7, %xmm7
-        cvtss2sd  (%rdi), %xmm7
-        cmpl      $74, %edx
+        cmpl      $970, %edx
         jle       ..B1.15
 ..B1.3:
-        movsd     1080+_vmldExpHATab(%rip), %xmm0
-        comisd    %xmm7, %xmm0
+        movsd     1080+_imldExpHATab(%rip), %xmm0
+        comisd    %xmm6, %xmm0
         jb        ..B1.14
 ..B1.4:
-        comisd    1096+_vmldExpHATab(%rip), %xmm7
+        comisd    1096+_imldExpHATab(%rip), %xmm6
         jb        ..B1.13
 ..B1.5:
-        movsd     1024+_vmldExpHATab(%rip), %xmm1
-        movaps    %xmm7, %xmm6
-        mulsd     %xmm7, %xmm1
-        lea       _vmldExpHATab(%rip), %r10
+        movsd     1024+_imldExpHATab(%rip), %xmm0
+        movaps    %xmm6, %xmm5
+        mulsd     %xmm6, %xmm0
+        lea       _imldExpHATab(%rip), %r10
+        movsd     %xmm0, -24(%rsp)
+        movsd     -24(%rsp), %xmm1
+        movq      1136+_imldExpHATab(%rip), %rdx
+        movq      %rdx, -8(%rsp)
+        addsd     1032+_imldExpHATab(%rip), %xmm1
         movsd     %xmm1, -16(%rsp)
         movsd     -16(%rsp), %xmm2
-        movsd     1072+_vmldExpHATab(%rip), %xmm1
-        movsd     1136+_vmldExpHATab(%rip), %xmm0
-        movsd     %xmm0, -24(%rsp)
-        addsd     1032+_vmldExpHATab(%rip), %xmm2
-        movsd     %xmm2, -8(%rsp)
-        movsd     -8(%rsp), %xmm3
-        movl      -8(%rsp), %edi
-        movl      %edi, %ecx
-        andl      $63, %edi
-        subsd     1032+_vmldExpHATab(%rip), %xmm3
-        movsd     %xmm3, -16(%rsp)
-        lea       1(%rdi,%rdi), %r8d
-        movsd     -16(%rsp), %xmm4
-        lea       (%rdi,%rdi), %r9d
-        mulsd     1104+_vmldExpHATab(%rip), %xmm4
-        movsd     -16(%rsp), %xmm5
-        subsd     %xmm4, %xmm6
-        mulsd     1112+_vmldExpHATab(%rip), %xmm5
+        movl      -16(%rsp), %r8d
+        movl      %r8d, %ecx
+        andl      $63, %r8d
+        subsd     1032+_imldExpHATab(%rip), %xmm2
+        movsd     %xmm2, -24(%rsp)
+        lea       1(%r8,%r8), %r9d
+        movsd     -24(%rsp), %xmm3
+        lea       (%r8,%r8), %edi
+        mulsd     1104+_imldExpHATab(%rip), %xmm3
+        movsd     -24(%rsp), %xmm4
+        subsd     %xmm3, %xmm5
+        mulsd     1112+_imldExpHATab(%rip), %xmm4
+        movsd     1072+_imldExpHATab(%rip), %xmm2
+        subsd     %xmm4, %xmm5
+        mulsd     %xmm5, %xmm2
         shrl      $6, %ecx
-        subsd     %xmm5, %xmm6
-        comisd    1088+_vmldExpHATab(%rip), %xmm7
-        mulsd     %xmm6, %xmm1
-        movsd     (%r10,%r9,8), %xmm0
+        addsd     1064+_imldExpHATab(%rip), %xmm2
+        comisd    1088+_imldExpHATab(%rip), %xmm6
+        mulsd     %xmm5, %xmm2
+        movsd     (%r10,%rdi,8), %xmm0
         lea       1023(%rcx), %edx
-        addsd     1064+_vmldExpHATab(%rip), %xmm1
-        mulsd     %xmm6, %xmm1
-        addsd     1056+_vmldExpHATab(%rip), %xmm1
-        mulsd     %xmm6, %xmm1
-        addsd     1048+_vmldExpHATab(%rip), %xmm1
-        mulsd     %xmm6, %xmm1
-        addsd     1040+_vmldExpHATab(%rip), %xmm1
-        mulsd     %xmm6, %xmm1
-        mulsd     %xmm6, %xmm1
-        addsd     %xmm6, %xmm1
-        addsd     (%r10,%r8,8), %xmm1
-        mulsd     %xmm0, %xmm1
+        addsd     1056+_imldExpHATab(%rip), %xmm2
+        mulsd     %xmm5, %xmm2
+        addsd     1048+_imldExpHATab(%rip), %xmm2
+        mulsd     %xmm5, %xmm2
+        addsd     1040+_imldExpHATab(%rip), %xmm2
+        mulsd     %xmm5, %xmm2
+        mulsd     %xmm5, %xmm2
+        addsd     %xmm5, %xmm2
+        addsd     (%r10,%r9,8), %xmm2
+        mulsd     %xmm0, %xmm2
         jb        ..B1.9
 ..B1.6:
         andl      $2047, %edx
-        addsd     %xmm0, %xmm1
+        addsd     %xmm0, %xmm2
         cmpl      $2046, %edx
         ja        ..B1.8
 ..B1.7:
-        movzwl    1142+_vmldExpHATab(%rip), %ecx
+        movq      1136+_imldExpHATab(%rip), %rcx
+        shrq      $48, %rcx
         shll      $4, %edx
         andl      $-32753, %ecx
         orl       %edx, %ecx
-        movw      %cx, -18(%rsp)
-        movsd     -24(%rsp), %xmm0
-        mulsd     %xmm0, %xmm1
-        cvtsd2ss  %xmm1, %xmm1
-        movss     %xmm1, (%rsi)
+        movw      %cx, -2(%rsp)
+        movsd     -8(%rsp), %xmm0
+        mulsd     %xmm0, %xmm2
+        movsd     %xmm2, (%rsi)
         ret
 ..B1.8:
         decl      %edx
         andl      $2047, %edx
-        movzwl    -18(%rsp), %ecx
+        movzwl    -2(%rsp), %ecx
         shll      $4, %edx
         andl      $-32753, %ecx
         orl       %edx, %ecx
-        movw      %cx, -18(%rsp)
-        movsd     -24(%rsp), %xmm0
-        mulsd     %xmm0, %xmm1
-        mulsd     1152+_vmldExpHATab(%rip), %xmm1
-        cvtsd2ss  %xmm1, %xmm1
-        movss     %xmm1, (%rsi)
+        movw      %cx, -2(%rsp)
+        movsd     -8(%rsp), %xmm0
+        mulsd     %xmm0, %xmm2
+        mulsd     1152+_imldExpHATab(%rip), %xmm2
+        movsd     %xmm2, (%rsi)
         ret
 ..B1.9:
         addl      $1083, %ecx
         andl      $2047, %ecx
         movl      %ecx, %eax
-        movzwl    -18(%rsp), %edx
+        movzwl    -2(%rsp), %edx
         shll      $4, %eax
         andl      $-32753, %edx
         orl       %eax, %edx
-        movw      %dx, -18(%rsp)
-        movsd     -24(%rsp), %xmm3
-        mulsd     %xmm3, %xmm1
-        mulsd     %xmm0, %xmm3
-        movaps    %xmm1, %xmm2
-        addsd     %xmm3, %xmm2
+        movw      %dx, -2(%rsp)
+        movsd     -8(%rsp), %xmm1
+        mulsd     %xmm1, %xmm2
+        mulsd     %xmm0, %xmm1
+        movaps    %xmm1, %xmm0
+        addsd     %xmm2, %xmm0
         cmpl      $50, %ecx
         ja        ..B1.11
 ..B1.10:
-        movsd     1160+_vmldExpHATab(%rip), %xmm0
-        mulsd     %xmm0, %xmm2
-        cvtsd2ss  %xmm2, %xmm2
+        mulsd     1160+_imldExpHATab(%rip), %xmm0
+        movsd     %xmm0, (%rsi)
         jmp       ..B1.12
 ..B1.11:
-        movsd     %xmm2, -72(%rsp)
+        movsd     %xmm0, -72(%rsp)
         movsd     -72(%rsp), %xmm0
-        subsd     %xmm0, %xmm3
-        movsd     %xmm3, -64(%rsp)
-        movsd     -64(%rsp), %xmm2
-        addsd     %xmm1, %xmm2
-        movsd     %xmm2, -64(%rsp)
-        movsd     -72(%rsp), %xmm1
-        mulsd     1168+_vmldExpHATab(%rip), %xmm1
-        movsd     %xmm1, -56(%rsp)
+        subsd     %xmm0, %xmm1
+        movsd     %xmm1, -64(%rsp)
+        movsd     -64(%rsp), %xmm1
+        addsd     %xmm2, %xmm1
+        movsd     %xmm1, -64(%rsp)
+        movsd     -72(%rsp), %xmm2
+        mulsd     1168+_imldExpHATab(%rip), %xmm2
+        movsd     %xmm2, -56(%rsp)
         movsd     -72(%rsp), %xmm4
         movsd     -56(%rsp), %xmm3
         addsd     %xmm3, %xmm4
@@ -184,69 +173,61 @@ __svml_sexp_ha_cout_rare_internal:
         addsd     %xmm9, %xmm10
         movsd     %xmm10, -32(%rsp)
         movsd     -40(%rsp), %xmm11
-        mulsd     1160+_vmldExpHATab(%rip), %xmm11
+        mulsd     1160+_imldExpHATab(%rip), %xmm11
         movsd     %xmm11, -40(%rsp)
         movsd     -32(%rsp), %xmm12
-        mulsd     1160+_vmldExpHATab(%rip), %xmm12
+        mulsd     1160+_imldExpHATab(%rip), %xmm12
         movsd     %xmm12, -32(%rsp)
-        movsd     -40(%rsp), %xmm2
+        movsd     -40(%rsp), %xmm14
         movsd     -32(%rsp), %xmm13
-        addsd     %xmm13, %xmm2
-        cvtsd2ss  %xmm2, %xmm2
+        addsd     %xmm13, %xmm14
+        movsd     %xmm14, (%rsi)
 ..B1.12:
-        movss     %xmm2, (%rsi)
         movl      $4, %eax
         ret
 ..B1.13:
-        movsd     1120+_vmldExpHATab(%rip), %xmm0
+        movsd     1120+_imldExpHATab(%rip), %xmm0
         movl      $4, %eax
         mulsd     %xmm0, %xmm0
-        cvtsd2ss  %xmm0, %xmm0
-        movss     %xmm0, (%rsi)
+        movsd     %xmm0, (%rsi)
         ret
 ..B1.14:
-        movsd     1128+_vmldExpHATab(%rip), %xmm0
+        movsd     1128+_imldExpHATab(%rip), %xmm0
         movl      $3, %eax
         mulsd     %xmm0, %xmm0
-        cvtsd2ss  %xmm0, %xmm0
-        movss     %xmm0, (%rsi)
+        movsd     %xmm0, (%rsi)
         ret
 ..B1.15:
-        movsd     1144+_vmldExpHATab(%rip), %xmm0
-        addsd     %xmm7, %xmm0
-        cvtsd2ss  %xmm0, %xmm0
-        movss     %xmm0, (%rsi)
+        movsd     1144+_imldExpHATab(%rip), %xmm0
+        addsd     %xmm6, %xmm0
+        movsd     %xmm0, (%rsi)
 ..B1.16:
         ret
 ..B1.17:
-        movb      3(%rdi), %dl
+        movb      -1(%rsp), %dl
         andb      $-128, %dl
         cmpb      $-128, %dl
         je        ..B1.19
 ..B1.18:
-        movss     (%rdi), %xmm0
-        mulss     %xmm0, %xmm0
-        movss     %xmm0, (%rsi)
+        mulsd     %xmm6, %xmm6
+        movsd     %xmm6, (%rsi)
         ret
 ..B1.19:
-        testl     $8388607, (%rdi)
+        testl     $1048575, -4(%rsp)
         jne       ..B1.18
 ..B1.20:
-        movsd     1136+_vmldExpHATab(%rip), %xmm0
-        cvtsd2ss  %xmm0, %xmm0
-        movss     %xmm0, (%rsi)
+        cmpl      $0, -8(%rsp)
+        jne       ..B1.18
+..B1.21:
+        movq      1136+_imldExpHATab(%rip), %rdx
+        movq      %rdx, (%rsi)
         ret
-        .align    16,0x90
-	.cfi_endproc
-	.type	__svml_sexp_ha_cout_rare_internal,@function
-	.size	__svml_sexp_ha_cout_rare_internal,.-__svml_sexp_ha_cout_rare_internal
-..LN__svml_sexp_ha_cout_rare_internal.0:
-	.data
-# -- End  __svml_sexp_ha_cout_rare_internal
+END(__avx2_exp_cout_rare_internal)
+# -- End  __avx2_exp_cout_rare_internal
 	.section .rodata, "a"
 	.align 32
 	.align 32
-_vmldExpHATab:
+_imldExpHATab:
 	.long	0
 	.long	1072693248
 	.long	0
@@ -517,12 +498,12 @@ _vmldExpHATab:
 	.long	1065423122
 	.long	472530941
 	.long	1062650218
-	.long	3758096384
-	.long	1079389762
-	.long	3758096384
-	.long	3226850697
-	.long	2147483648
-	.long	3227123254
+	.long	4277811695
+	.long	1082535490
+	.long	3715808466
+	.long	3230016299
+	.long	3576508497
+	.long	3230091536
 	.long	4277796864
 	.long	1065758274
 	.long	3164486458
@@ -541,8 +522,9 @@ _vmldExpHATab:
 	.long	1009778688
 	.long	0
 	.long	1106771968
-	.type	_vmldExpHATab,@object
-	.size	_vmldExpHATab,1176
+	.type	_imldExpHATab,@object
+	.size	_imldExpHATab,1176
 	.data
 	.section .note.GNU-stack, ""
 # End
+
diff --git a/libm/x86_64/exp_data_avx2.S b/libm/x86_64/exp_data_avx2.S
index f9a98e8..1affdf9 100644
--- a/libm/x86_64/exp_data_avx2.S
+++ b/libm/x86_64/exp_data_avx2.S
@@ -30,863 +30,473 @@ SUCH DAMAGE.
 	.section .rodata, "a"
 	.align 64
 	.align 64
-	.hidden __svml_sexp_ha_data_internal_avx512
-	.globl __svml_sexp_ha_data_internal_avx512
-__svml_sexp_ha_data_internal_avx512:
-	.long	0
-	.long	3007986186
-	.long	860277610
-	.long	3010384254
-	.long	2991457809
-	.long	3008462297
-	.long	860562562
-	.long	3004532446
-	.long	856238081
-	.long	3001480295
-	.long	857441778
-	.long	815380209
-	.long	3003456168
-	.long	3001196762
-	.long	2986372182
-	.long	3006683458
-	.long	848495278
-	.long	851809756
-	.long	3003311522
-	.long	2995654817
-	.long	833868005
-	.long	3004843819
-	.long	835836658
-	.long	3003498340
-	.long	2994528642
-	.long	3002229827
-	.long	2981408986
-	.long	2983889551
-	.long	2983366846
-	.long	3000350873
-	.long	833659207
-	.long	2987748092
-	.long	1065353216
-	.long	1065536903
-	.long	1065724611
-	.long	1065916431
-	.long	1066112450
-	.long	1066312762
-	.long	1066517459
-	.long	1066726640
-	.long	1066940400
-	.long	1067158842
-	.long	1067382066
-	.long	1067610179
-	.long	1067843287
-	.long	1068081499
-	.long	1068324927
-	.long	1068573686
-	.long	1068827891
-	.long	1069087663
-	.long	1069353124
-	.long	1069624397
-	.long	1069901610
-	.long	1070184894
-	.long	1070474380
-	.long	1070770206
-	.long	1071072509
-	.long	1071381432
-	.long	1071697119
-	.long	1072019719
-	.long	1072349383
-	.long	1072686266
-	.long	1073030525
-	.long	1073382323
-	.long	1069066811
-	.long	1069066811
-	.long	1069066811
-	.long	1069066811
-	.long	1069066811
-	.long	1069066811
-	.long	1069066811
-	.long	1069066811
-	.long	1069066811
-	.long	1069066811
-	.long	1069066811
-	.long	1069066811
-	.long	1069066811
-	.long	1069066811
-	.long	1069066811
-	.long	1069066811
-	.long	1220542464
-	.long	1220542464
-	.long	1220542464
-	.long	1220542464
-	.long	1220542464
-	.long	1220542464
-	.long	1220542464
-	.long	1220542464
-	.long	1220542464
-	.long	1220542464
-	.long	1220542464
-	.long	1220542464
-	.long	1220542464
-	.long	1220542464
-	.long	1220542464
-	.long	1220542464
-	.long	1060205080
-	.long	1060205080
-	.long	1060205080
-	.long	1060205080
-	.long	1060205080
-	.long	1060205080
-	.long	1060205080
-	.long	1060205080
-	.long	1060205080
-	.long	1060205080
-	.long	1060205080
-	.long	1060205080
-	.long	1060205080
-	.long	1060205080
-	.long	1060205080
-	.long	1060205080
-	.long	2969756424
-	.long	2969756424
-	.long	2969756424
-	.long	2969756424
-	.long	2969756424
-	.long	2969756424
-	.long	2969756424
-	.long	2969756424
-	.long	2969756424
-	.long	2969756424
-	.long	2969756424
-	.long	2969756424
-	.long	2969756424
-	.long	2969756424
-	.long	2969756424
-	.long	2969756424
-	.long	3221225471
-	.long	3221225471
-	.long	3221225471
-	.long	3221225471
-	.long	3221225471
-	.long	3221225471
-	.long	3221225471
-	.long	3221225471
-	.long	3221225471
-	.long	3221225471
-	.long	3221225471
-	.long	3221225471
-	.long	3221225471
-	.long	3221225471
-	.long	3221225471
-	.long	3221225471
-	.long	1042983923
-	.long	1042983923
-	.long	1042983923
-	.long	1042983923
-	.long	1042983923
-	.long	1042983923
-	.long	1042983923
-	.long	1042983923
-	.long	1042983923
-	.long	1042983923
-	.long	1042983923
-	.long	1042983923
-	.long	1042983923
-	.long	1042983923
-	.long	1042983923
-	.long	1042983923
-	.long	1056964854
-	.long	1056964854
-	.long	1056964854
-	.long	1056964854
-	.long	1056964854
-	.long	1056964854
-	.long	1056964854
-	.long	1056964854
-	.long	1056964854
-	.long	1056964854
-	.long	1056964854
-	.long	1056964854
-	.long	1056964854
-	.long	1056964854
-	.long	1056964854
-	.long	1056964854
-	.long	2147483647
-	.long	2147483647
-	.long	2147483647
-	.long	2147483647
-	.long	2147483647
-	.long	2147483647
-	.long	2147483647
-	.long	2147483647
-	.long	2147483647
-	.long	2147483647
-	.long	2147483647
-	.long	2147483647
-	.long	2147483647
-	.long	2147483647
-	.long	2147483647
-	.long	2147483647
-	.long	1118743631
-	.long	1118743631
-	.long	1118743631
-	.long	1118743631
-	.long	1118743631
-	.long	1118743631
-	.long	1118743631
-	.long	1118743631
-	.long	1118743631
-	.long	1118743631
-	.long	1118743631
-	.long	1118743631
-	.long	1118743631
-	.long	1118743631
-	.long	1118743631
-	.long	1118743631
-	.long	796917760
-	.long	796917760
-	.long	796917760
-	.long	796917760
-	.long	796917760
-	.long	796917760
-	.long	796917760
-	.long	796917760
-	.long	796917760
-	.long	796917760
-	.long	796917760
-	.long	796917760
-	.long	796917760
-	.long	796917760
-	.long	796917760
-	.long	796917760
-	.long	124
-	.long	124
-	.long	124
-	.long	124
-	.long	124
-	.long	124
-	.long	124
-	.long	124
-	.long	124
-	.long	124
-	.long	124
-	.long	124
-	.long	124
-	.long	124
-	.long	124
-	.long	124
-	.type	__svml_sexp_ha_data_internal_avx512,@object
-	.size	__svml_sexp_ha_data_internal_avx512,960
-	.align 64
-	.hidden __svml_sexp_ha_data_internal
-	.globl __svml_sexp_ha_data_internal
-__svml_sexp_ha_data_internal:
-	.long	1085844027
-	.long	1085844027
-	.long	1085844027
-	.long	1085844027
-	.long	1085844027
-	.long	1085844027
-	.long	1085844027
-	.long	1085844027
-	.long	1085844027
-	.long	1085844027
-	.long	1085844027
-	.long	1085844027
-	.long	1085844027
-	.long	1085844027
-	.long	1085844027
-	.long	1085844027
-	.long	1262486012
-	.long	1262486012
-	.long	1262486012
-	.long	1262486012
-	.long	1262486012
-	.long	1262486012
-	.long	1262486012
-	.long	1262486012
-	.long	1262486012
-	.long	1262486012
-	.long	1262486012
-	.long	1262486012
-	.long	1262486012
-	.long	1262486012
-	.long	1262486012
-	.long	1262486012
-	.long	2147483647
-	.long	2147483647
-	.long	2147483647
-	.long	2147483647
-	.long	2147483647
-	.long	2147483647
-	.long	2147483647
-	.long	2147483647
+	.hidden __avx2_exp_data_internal
+	.globl __avx2_exp_data_internal
+__avx2_exp_data_internal:
+	.long	0
+	.long	1072693248
+	.long	0
+	.long	0
+	.long	1048019041
+	.long	1072704666
+	.long	1398474845
+	.long	3161559171
+	.long	3541402996
+	.long	1072716208
+	.long	2759177317
+	.long	1015903202
+	.long	410360776
+	.long	1072727877
+	.long	1269990655
+	.long	1013024446
+	.long	1828292879
+	.long	1072739672
+	.long	1255956747
+	.long	1016636974
+	.long	852742562
+	.long	1072751596
+	.long	667253587
+	.long	1010842135
+	.long	3490863953
+	.long	1072763649
+	.long	960797498
+	.long	3163997456
+	.long	2930322912
+	.long	1072775834
+	.long	2599499422
+	.long	3163762623
+	.long	1014845819
+	.long	1072788152
+	.long	3117910646
+	.long	3162607681
+	.long	3949972341
+	.long	1072800603
+	.long	2068408548
+	.long	1015962444
+	.long	828946858
+	.long	1072813191
+	.long	10642492
+	.long	1016988014
+	.long	2288159958
+	.long	1072825915
+	.long	2169144469
+	.long	1015924597
+	.long	1853186616
+	.long	1072838778
+	.long	3066496371
+	.long	1016705150
+	.long	1709341917
+	.long	1072851781
+	.long	2571168217
+	.long	1015201075
+	.long	4112506593
+	.long	1072864925
+	.long	2947355221
+	.long	1015419624
+	.long	2799960843
+	.long	1072878213
+	.long	1423655381
+	.long	1016070727
+	.long	171030293
+	.long	1072891646
+	.long	3526460132
+	.long	1015477354
+	.long	2992903935
+	.long	1072905224
+	.long	2218154406
+	.long	1016276769
+	.long	926591435
+	.long	1072918951
+	.long	3208833762
+	.long	3163962090
+	.long	887463927
+	.long	1072932827
+	.long	3596744163
+	.long	3161842742
+	.long	1276261410
+	.long	1072946854
+	.long	300981948
+	.long	1015732745
+	.long	569847338
+	.long	1072961034
+	.long	472945272
+	.long	3160339305
+	.long	1617004845
+	.long	1072975368
+	.long	82804944
+	.long	1011391354
+	.long	3049340112
+	.long	1072989858
+	.long	3062915824
+	.long	1014219171
+	.long	3577096743
+	.long	1073004506
+	.long	2951496418
+	.long	1014842263
+	.long	1990012071
+	.long	1073019314
+	.long	3529070563
+	.long	3163861769
+	.long	1453150082
+	.long	1073034283
+	.long	498154669
+	.long	3162536638
+	.long	917841882
+	.long	1073049415
+	.long	18715565
+	.long	1016707884
+	.long	3712504873
+	.long	1073064711
+	.long	88491949
+	.long	1016476236
+	.long	363667784
+	.long	1073080175
+	.long	813753950
+	.long	1016833785
+	.long	2956612997
+	.long	1073095806
+	.long	2118169751
+	.long	3163784129
+	.long	2186617381
+	.long	1073111608
+	.long	2270764084
+	.long	3164321289
+	.long	1719614413
+	.long	1073127582
+	.long	330458198
+	.long	3164331316
+	.long	1013258799
+	.long	1073143730
+	.long	1748797611
+	.long	3161177658
+	.long	3907805044
+	.long	1073160053
+	.long	2257091225
+	.long	3162598983
+	.long	1447192521
+	.long	1073176555
+	.long	1462857171
+	.long	3163563097
+	.long	1944781191
+	.long	1073193236
+	.long	3993278767
+	.long	3162772855
+	.long	919555682
+	.long	1073210099
+	.long	3121969534
+	.long	1013996802
+	.long	2571947539
+	.long	1073227145
+	.long	3558159064
+	.long	3164425245
+	.long	2604962541
+	.long	1073244377
+	.long	2614425274
+	.long	3164587768
+	.long	1110089947
+	.long	1073261797
+	.long	1451641639
+	.long	1016523249
+	.long	2568320822
+	.long	1073279406
+	.long	2732824428
+	.long	1015401491
+	.long	2966275557
+	.long	1073297207
+	.long	2176155324
+	.long	3160891335
+	.long	2682146384
+	.long	1073315202
+	.long	2082178513
+	.long	3164411995
+	.long	2191782032
+	.long	1073333393
+	.long	2960257726
+	.long	1014791238
+	.long	2069751141
+	.long	1073351782
+	.long	1562170675
+	.long	3163773257
+	.long	2990417245
+	.long	1073370371
+	.long	3683467745
+	.long	3164417902
+	.long	1434058175
+	.long	1073389163
+	.long	251133233
+	.long	1016134345
+	.long	2572866477
+	.long	1073408159
+	.long	878562433
+	.long	1016570317
+	.long	3092190715
+	.long	1073427362
+	.long	814012168
+	.long	3160571998
+	.long	4076559943
+	.long	1073446774
+	.long	2119478331
+	.long	3161806927
+	.long	2420883922
+	.long	1073466398
+	.long	2049810052
+	.long	1015168464
+	.long	3716502172
+	.long	1073486235
+	.long	2303740125
+	.long	1015091301
+	.long	777507147
+	.long	1073506289
+	.long	4282924205
+	.long	1016236109
+	.long	3706687593
+	.long	1073526560
+	.long	3521726939
+	.long	1014301643
+	.long	1242007932
+	.long	1073547053
+	.long	1132034716
+	.long	3164388407
+	.long	3707479175
+	.long	1073567768
+	.long	3613079303
+	.long	1015213314
+	.long	64696965
+	.long	1073588710
+	.long	1768797490
+	.long	1016865536
+	.long	863738719
+	.long	1073609879
+	.long	1326992220
+	.long	3163661773
+	.long	3884662774
+	.long	1073631278
+	.long	2158611599
+	.long	1015258761
+	.long	2728693978
+	.long	1073652911
+	.long	396109971
+	.long	3164511267
+	.long	3999357479
+	.long	1073674779
+	.long	2258941616
+	.long	1016973300
+	.long	1533953344
+	.long	1073696886
+	.long	769171851
+	.long	1016714209
+	.long	2174652632
+	.long	1073719233
+	.long	4087714590
+	.long	1015498835
+	.long	1697350398
+	.long	1079448903
+	.long	1697350398
+	.long	1079448903
+	.long	1697350398
+	.long	1079448903
+	.long	1697350398
+	.long	1079448903
+	.long	1697350398
+	.long	1079448903
+	.long	1697350398
+	.long	1079448903
+	.long	1697350398
+	.long	1079448903
+	.long	1697350398
+	.long	1079448903
+	.long	0
+	.long	1127743488
+	.long	0
+	.long	1127743488
+	.long	0
+	.long	1127743488
+	.long	0
+	.long	1127743488
+	.long	0
+	.long	1127743488
+	.long	0
+	.long	1127743488
+	.long	0
+	.long	1127743488
+	.long	0
+	.long	1127743488
+	.long	0
+	.long	1123549184
+	.long	0
+	.long	1123549184
+	.long	0
+	.long	1123549184
+	.long	0
+	.long	1123549184
+	.long	0
+	.long	1123549184
+	.long	0
+	.long	1123549184
+	.long	0
+	.long	1123549184
+	.long	0
+	.long	1123549184
+	.long	4277796864
+	.long	1065758274
+	.long	4277796864
+	.long	1065758274
+	.long	4277796864
+	.long	1065758274
+	.long	4277796864
+	.long	1065758274
+	.long	4277796864
+	.long	1065758274
+	.long	4277796864
+	.long	1065758274
+	.long	4277796864
+	.long	1065758274
+	.long	4277796864
+	.long	1065758274
+	.long	3164486458
+	.long	1025308570
+	.long	3164486458
+	.long	1025308570
+	.long	3164486458
+	.long	1025308570
+	.long	3164486458
+	.long	1025308570
+	.long	3164486458
+	.long	1025308570
+	.long	3164486458
+	.long	1025308570
+	.long	3164486458
+	.long	1025308570
+	.long	3164486458
+	.long	1025308570
+	.long	4294957883
+	.long	1071644671
+	.long	4294957883
+	.long	1071644671
+	.long	4294957883
+	.long	1071644671
+	.long	4294957883
+	.long	1071644671
+	.long	4294957883
+	.long	1071644671
+	.long	4294957883
+	.long	1071644671
+	.long	4294957883
+	.long	1071644671
+	.long	4294957883
+	.long	1071644671
+	.long	1431659838
+	.long	1069897045
+	.long	1431659838
+	.long	1069897045
+	.long	1431659838
+	.long	1069897045
+	.long	1431659838
+	.long	1069897045
+	.long	1431659838
+	.long	1069897045
+	.long	1431659838
+	.long	1069897045
+	.long	1431659838
+	.long	1069897045
+	.long	1431659838
+	.long	1069897045
+	.long	1059163027
+	.long	1067799895
+	.long	1059163027
+	.long	1067799895
+	.long	1059163027
+	.long	1067799895
+	.long	1059163027
+	.long	1067799895
+	.long	1059163027
+	.long	1067799895
+	.long	1059163027
+	.long	1067799895
+	.long	1059163027
+	.long	1067799895
+	.long	1059163027
+	.long	1067799895
+	.long	765416603
+	.long	1065423121
+	.long	765416603
+	.long	1065423121
+	.long	765416603
+	.long	1065423121
+	.long	765416603
+	.long	1065423121
+	.long	765416603
+	.long	1065423121
+	.long	765416603
+	.long	1065423121
+	.long	765416603
+	.long	1065423121
+	.long	765416603
+	.long	1065423121
+	.long	63
+	.long	0
+	.long	63
+	.long	0
+	.long	63
+	.long	0
+	.long	63
+	.long	0
+	.long	63
+	.long	0
+	.long	63
+	.long	0
+	.long	63
+	.long	0
+	.long	63
+	.long	0
+	.long	1023
+	.long	0
+	.long	1023
+	.long	0
+	.long	1023
+	.long	0
+	.long	1023
+	.long	0
+	.long	1023
+	.long	0
+	.long	1023
+	.long	0
+	.long	1023
+	.long	0
+	.long	1023
+	.long	0
+	.long	4294967295
 	.long	2147483647
+	.long	4294967295
 	.long	2147483647
+	.long	4294967295
 	.long	2147483647
+	.long	4294967295
 	.long	2147483647
+	.long	4294967295
 	.long	2147483647
+	.long	4294967295
 	.long	2147483647
+	.long	4294967295
 	.long	2147483647
+	.long	4294967295
 	.long	2147483647
-	.long	1118743631
-	.long	1118743631
-	.long	1118743631
-	.long	1118743631
-	.long	1118743631
-	.long	1118743631
-	.long	1118743631
-	.long	1118743631
-	.long	1118743631
-	.long	1118743631
-	.long	1118743631
-	.long	1118743631
-	.long	1118743631
-	.long	1118743631
-	.long	1118743631
-	.long	1118743631
-	.long	1043427328
-	.long	1043427328
-	.long	1043427328
-	.long	1043427328
-	.long	1043427328
-	.long	1043427328
-	.long	1043427328
-	.long	1043427328
-	.long	1043427328
-	.long	1043427328
-	.long	1043427328
-	.long	1043427328
-	.long	1043427328
-	.long	1043427328
-	.long	1043427328
-	.long	1043427328
-	.long	923139572
-	.long	923139572
-	.long	923139572
-	.long	923139572
-	.long	923139572
-	.long	923139572
-	.long	923139572
-	.long	923139572
-	.long	923139572
-	.long	923139572
-	.long	923139572
-	.long	923139572
-	.long	923139572
-	.long	923139572
-	.long	923139572
-	.long	923139572
-	.long	0
-	.long	856238081
-	.long	848495278
-	.long	2994528642
-	.long	0
-	.long	856238081
-	.long	848495278
-	.long	2994528642
-	.long	0
-	.long	856238081
-	.long	848495278
-	.long	2994528642
-	.long	0
-	.long	856238081
-	.long	848495278
-	.long	2994528642
-	.long	1065353216
-	.long	1066940400
-	.long	1068827891
-	.long	1071072509
-	.long	1065353216
-	.long	1066940400
-	.long	1068827891
-	.long	1071072509
-	.long	1065353216
-	.long	1066940400
-	.long	1068827891
-	.long	1071072509
-	.long	1065353216
-	.long	1066940400
-	.long	1068827891
-	.long	1071072509
-	.long	1026210578
-	.long	1026210578
-	.long	1026210578
-	.long	1026210578
-	.long	1026210578
-	.long	1026210578
-	.long	1026210578
-	.long	1026210578
-	.long	1026210578
-	.long	1026210578
-	.long	1026210578
-	.long	1026210578
-	.long	1026210578
-	.long	1026210578
-	.long	1026210578
-	.long	1026210578
-	.long	1042987794
-	.long	1042987794
-	.long	1042987794
-	.long	1042987794
-	.long	1042987794
-	.long	1042987794
-	.long	1042987794
-	.long	1042987794
-	.long	1042987794
-	.long	1042987794
-	.long	1042987794
-	.long	1042987794
-	.long	1042987794
-	.long	1042987794
-	.long	1042987794
-	.long	1042987794
-	.long	1056964607
-	.long	1056964607
-	.long	1056964607
-	.long	1056964607
-	.long	1056964607
-	.long	1056964607
-	.long	1056964607
-	.long	1056964607
-	.long	1056964607
-	.long	1056964607
-	.long	1056964607
-	.long	1056964607
-	.long	1056964607
-	.long	1056964607
-	.long	1056964607
-	.long	1056964607
-	.long	2139095040
-	.long	2139095040
-	.long	2139095040
-	.long	2139095040
-	.long	2139095040
-	.long	2139095040
-	.long	2139095040
-	.long	2139095040
-	.long	2139095040
-	.long	2139095040
-	.long	2139095040
-	.long	2139095040
-	.long	2139095040
-	.long	2139095040
-	.long	2139095040
-	.long	2139095040
-	.long	1065353216
-	.long	0
-	.long	1065398765
-	.long	857715507
-	.long	1065444562
-	.long	2982428624
-	.long	1065490607
-	.long	856090616
-	.long	1065536903
-	.long	3008276393
-	.long	1065583449
-	.long	857595758
-	.long	1065630249
-	.long	3002158215
-	.long	1065677302
-	.long	858550072
-	.long	1065724611
-	.long	860854417
-	.long	1065772177
-	.long	859251839
-	.long	1065820002
-	.long	3008545510
-	.long	1065868085
-	.long	864021810
-	.long	1065916431
-	.long	3011435083
-	.long	1065965038
-	.long	819487640
-	.long	1066013909
-	.long	862622430
-	.long	1066063046
-	.long	856316133
-	.long	1066112450
-	.long	2992679841
-	.long	1066162122
-	.long	3001970243
-	.long	1066212063
-	.long	859241655
-	.long	1066262276
-	.long	847285146
-	.long	1066312762
-	.long	3010032741
-	.long	1066363521
-	.long	843003463
-	.long	1066414556
-	.long	856041677
-	.long	1066465869
-	.long	3009946167
-	.long	1066517459
-	.long	862410276
-	.long	1066569330
-	.long	863709796
-	.long	1066621483
-	.long	851201948
-	.long	1066673919
-	.long	838513088
-	.long	1066726640
-	.long	3006136850
-	.long	1066779647
-	.long	3010613863
-	.long	1066832941
-	.long	853311585
-	.long	1066886525
-	.long	856941301
-	.long	1066940400
-	.long	857938801
-	.long	1066994568
-	.long	2993474036
-	.long	1067049030
-	.long	3009003152
-	.long	1067103787
-	.long	2967126046
-	.long	1067158842
-	.long	3003929955
-	.long	1067214195
-	.long	860904838
-	.long	1067269850
-	.long	3006375425
-	.long	1067325806
-	.long	843377209
-	.long	1067382066
-	.long	859906882
-	.long	1067438632
-	.long	848662531
-	.long	1067495505
-	.long	838571797
-	.long	1067552687
-	.long	3000363757
-	.long	1067610179
-	.long	818090586
-	.long	1067667983
-	.long	860999685
-	.long	1067726102
-	.long	3005475891
-	.long	1067784536
-	.long	3010626242
-	.long	1067843287
-	.long	3006045534
-	.long	1067902357
-	.long	2998944862
-	.long	1067961748
-	.long	3004575030
-	.long	1068021461
-	.long	842258832
-	.long	1068081499
-	.long	3004574472
-	.long	1068141862
-	.long	858118433
-	.long	1068202554
-	.long	3004476802
-	.long	1068263575
-	.long	3006694272
-	.long	1068324927
-	.long	2989353718
-	.long	1068386612
-	.long	858100843
-	.long	1068448633
-	.long	3005558251
-	.long	1068510990
-	.long	3000050815
-	.long	1068573686
-	.long	3011271336
-	.long	1068636722
-	.long	3006477262
-	.long	1068700100
-	.long	840255625
-	.long	1068763823
-	.long	3007001961
-	.long	1068827891
-	.long	852486010
-	.long	1068892308
-	.long	3006218836
-	.long	1068957074
-	.long	2993076596
-	.long	1069022192
-	.long	3000356208
-	.long	1069087663
-	.long	856606199
-	.long	1069153490
-	.long	856315927
-	.long	1069219675
-	.long	3004946819
-	.long	1069286218
-	.long	863888852
-	.long	1069353124
-	.long	3007401960
-	.long	1069420392
-	.long	832069785
-	.long	1069488026
-	.long	3004369690
-	.long	1069556027
-	.long	3007061599
-	.long	1069624397
-	.long	3000395326
-	.long	1069693138
-	.long	851736822
-	.long	1069762253
-	.long	2996268450
-	.long	1069831743
-	.long	2999890733
-	.long	1069901610
-	.long	839559223
-	.long	1069971857
-	.long	3004476343
-	.long	1070042485
-	.long	3000848665
-	.long	1070113496
-	.long	859381756
-	.long	1070184894
-	.long	3010667426
-	.long	1070256678
-	.long	859604257
-	.long	1070328853
-	.long	2953234114
-	.long	1070401420
-	.long	3010682756
-	.long	1070474380
-	.long	841546788
-	.long	1070547737
-	.long	2999163804
-	.long	1070621492
-	.long	2996061011
-	.long	1070695647
-	.long	860136147
-	.long	1070770206
-	.long	3009158570
-	.long	1070845169
-	.long	3008165066
-	.long	1070920539
-	.long	3005767716
-	.long	1070996318
-	.long	2955948569
-	.long	1071072509
-	.long	3000280458
-	.long	1071149113
-	.long	857267046
-	.long	1071226134
-	.long	3002473793
-	.long	1071303572
-	.long	861820308
-	.long	1071381432
-	.long	3008383516
-	.long	1071459714
-	.long	3010850715
-	.long	1071538421
-	.long	3011199971
-	.long	1071617555
-	.long	2995067967
-	.long	1071697119
-	.long	2988354393
-	.long	1071777115
-	.long	2889750092
-	.long	1071857545
-	.long	862289173
-	.long	1071938413
-	.long	3007287685
-	.long	1072019719
-	.long	2990808038
-	.long	1072101467
-	.long	3005883328
-	.long	1072183658
-	.long	857766040
-	.long	1072266296
-	.long	855693471
-	.long	1072349383
-	.long	2990610142
-	.long	1072432921
-	.long	3004393415
-	.long	1072516912
-	.long	846646433
-	.long	1072601360
-	.long	3008357562
-	.long	1072686266
-	.long	3007858250
-	.long	1072771633
-	.long	3006309869
-	.long	1072857463
-	.long	854681599
-	.long	1072943760
-	.long	2995197552
-	.long	1073030525
-	.long	841557046
-	.long	1073117761
-	.long	861464191
-	.long	1073205471
-	.long	862328226
-	.long	1073293658
-	.long	3003728983
-	.long	1073382323
-	.long	2995926872
-	.long	1073471469
-	.long	861522913
-	.long	1073561100
-	.long	835668076
-	.long	1073651217
-	.long	859371295
-	.long	1127787067
-	.long	1127787067
-	.long	1127787067
-	.long	1127787067
-	.long	1127787067
-	.long	1127787067
-	.long	1127787067
-	.long	1127787067
-	.long	1127787067
-	.long	1127787067
-	.long	1127787067
-	.long	1127787067
-	.long	1127787067
-	.long	1127787067
-	.long	1127787067
-	.long	1127787067
-	.long	1262485504
-	.long	1262485504
-	.long	1262485504
-	.long	1262485504
-	.long	1262485504
-	.long	1262485504
-	.long	1262485504
-	.long	1262485504
-	.long	1262485504
-	.long	1262485504
-	.long	1262485504
-	.long	1262485504
-	.long	1262485504
-	.long	1262485504
-	.long	1262485504
-	.long	1262485504
-	.long	1001488384
-	.long	1001488384
-	.long	1001488384
-	.long	1001488384
-	.long	1001488384
-	.long	1001488384
-	.long	1001488384
-	.long	1001488384
-	.long	1001488384
-	.long	1001488384
-	.long	1001488384
-	.long	1001488384
-	.long	1001488384
-	.long	1001488384
-	.long	1001488384
-	.long	1001488384
-	.long	3051257987
-	.long	3051257987
-	.long	3051257987
-	.long	3051257987
-	.long	3051257987
-	.long	3051257987
-	.long	3051257987
-	.long	3051257987
-	.long	3051257987
-	.long	3051257987
-	.long	3051257987
-	.long	3051257987
-	.long	3051257987
-	.long	3051257987
-	.long	3051257987
-	.long	3051257987
-	.long	16256
-	.long	16256
-	.long	16256
-	.long	16256
-	.long	16256
-	.long	16256
-	.long	16256
-	.long	16256
-	.long	16256
-	.long	16256
-	.long	16256
-	.long	16256
-	.long	16256
-	.long	16256
-	.long	16256
-	.long	16256
-	.long	2139095040
-	.long	2139095040
-	.long	2139095040
-	.long	2139095040
-	.long	2139095040
-	.long	2139095040
-	.long	2139095040
-	.long	2139095040
-	.long	2139095040
-	.long	2139095040
-	.long	2139095040
-	.long	2139095040
-	.long	2139095040
-	.long	2139095040
-	.long	2139095040
-	.long	2139095040
-	.long	127
-	.long	127
-	.long	127
-	.long	127
-	.long	127
-	.long	127
-	.long	127
-	.long	127
-	.long	127
-	.long	127
-	.long	127
-	.long	127
-	.long	127
-	.long	127
-	.long	127
-	.long	127
-	.long	4294967232
-	.long	4294967232
-	.long	4294967232
-	.long	4294967232
-	.long	4294967232
-	.long	4294967232
-	.long	4294967232
-	.long	4294967232
-	.long	4294967232
-	.long	4294967232
-	.long	4294967232
-	.long	4294967232
-	.long	4294967232
-	.long	4294967232
-	.long	4294967232
-	.long	4294967232
-	.long	864026624
-	.long	864026624
-	.long	864026624
-	.long	864026624
-	.long	864026624
-	.long	864026624
-	.long	864026624
-	.long	864026624
-	.long	864026624
-	.long	864026624
-	.long	864026624
-	.long	864026624
-	.long	864026624
-	.long	864026624
-	.long	864026624
-	.long	864026624
-	.long	1056964608
-	.long	1056964608
-	.long	1056964608
-	.long	1056964608
-	.long	1056964608
-	.long	1056964608
-	.long	1056964608
-	.long	1056964608
-	.long	1056964608
-	.long	1056964608
-	.long	1056964608
-	.long	1056964608
-	.long	1056964608
-	.long	1056964608
-	.long	1056964608
-	.long	1056964608
+	.long	4294967295
+	.long	1082532650
+	.long	4294967295
+	.long	1082532650
+	.long	4294967295
+	.long	1082532650
+	.long	4294967295
+	.long	1082532650
+	.long	4294967295
+	.long	1082532650
+	.long	4294967295
+	.long	1082532650
+	.long	4294967295
+	.long	1082532650
+	.long	4294967295
+	.long	1082532650
 	.long	2147483647
 	.long	2147483647
 	.long	2147483647
@@ -903,56 +513,56 @@ __svml_sexp_ha_data_internal:
 	.long	2147483647
 	.long	2147483647
 	.long	2147483647
-	.long	1118743631
-	.long	1118743631
-	.long	1118743631
-	.long	1118743631
-	.long	1118743631
-	.long	1118743631
-	.long	1118743631
-	.long	1118743631
-	.long	1118743631
-	.long	1118743631
-	.long	1118743631
-	.long	1118743631
-	.long	1118743631
-	.long	1118743631
-	.long	1118743631
-	.long	1118743631
-	.long	1118925335
-	.long	1118925335
-	.long	1118925335
-	.long	1118925335
-	.long	1118925335
-	.long	1118925335
-	.long	1118925335
-	.long	1118925335
-	.long	1118925335
-	.long	1118925335
-	.long	1118925335
-	.long	1118925335
-	.long	1118925335
-	.long	1118925335
-	.long	1118925335
-	.long	1118925335
-	.long	3268407732
-	.long	3268407732
-	.long	3268407732
-	.long	3268407732
-	.long	3268407732
-	.long	3268407732
-	.long	3268407732
-	.long	3268407732
-	.long	3268407732
-	.long	3268407732
-	.long	3268407732
-	.long	3268407732
-	.long	3268407732
-	.long	3268407732
-	.long	3268407732
-	.long	3268407732
-	.type	__svml_sexp_ha_data_internal,@object
-	.size	__svml_sexp_ha_data_internal,2688
+	.long	1082532650
+	.long	1082532650
+	.long	1082532650
+	.long	1082532650
+	.long	1082532650
+	.long	1082532650
+	.long	1082532650
+	.long	1082532650
+	.long	1082532650
+	.long	1082532650
+	.long	1082532650
+	.long	1082532650
+	.long	1082532650
+	.long	1082532650
+	.long	1082532650
+	.long	1082532650
+	.long	4277811695
+	.long	1082535490
+	.long	4277811695
+	.long	1082535490
+	.long	4277811695
+	.long	1082535490
+	.long	4277811695
+	.long	1082535490
+	.long	4277811695
+	.long	1082535490
+	.long	4277811695
+	.long	1082535490
+	.long	4277811695
+	.long	1082535490
+	.long	4277811695
+	.long	1082535490
+	.long	3576508497
+	.long	3230091536
+	.long	3576508497
+	.long	3230091536
+	.long	3576508497
+	.long	3230091536
+	.long	3576508497
+	.long	3230091536
+	.long	3576508497
+	.long	3230091536
+	.long	3576508497
+	.long	3230091536
+	.long	3576508497
+	.long	3230091536
+	.long	3576508497
+	.long	3230091536
+	.type	__avx2_exp_data_internal,@object
+	.size	__avx2_exp_data_internal,2112
 	.data
 	.section .note.GNU-stack, ""
 # End
diff --git a/libm/x86_64/expf_cout_avx2.S b/libm/x86_64/expf_cout_avx2.S
new file mode 100644
index 0000000..2fd42c3
--- /dev/null
+++ b/libm/x86_64/expf_cout_avx2.S
@@ -0,0 +1,548 @@
+/*
+Copyright (C) 2019 The Android Open Source Project
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions
+are met:
+ * Redistributions of source code must retain the above copyright
+   notice, this list of conditions and the following disclaimer.
+ * Redistributions in binary form must reproduce the above copyright
+   notice, this list of conditions and the following disclaimer in
+   the documentation and/or other materials provided with the
+   distribution.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT
+OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+SUCH DAMAGE.
+*/
+
+# -- Machine type EM64t
+	.text
+..TXTST0:
+.L_2__routine_start___svml_sexp_ha_cout_rare_internal_0:
+# -- Begin  __svml_sexp_ha_cout_rare_internal
+	.text
+       .align    16,0x90
+	.hidden __svml_sexp_ha_cout_rare_internal
+	.globl __svml_sexp_ha_cout_rare_internal
+__svml_sexp_ha_cout_rare_internal:
+# parameter 1: %rdi
+# parameter 2: %rsi
+..B1.1:
+	.cfi_startproc
+..___tag_value___svml_sexp_ha_cout_rare_internal.1:
+..L2:
+
+        xorl      %eax, %eax
+        movzwl    2(%rdi), %edx
+        andl      $32640, %edx
+        shrl      $7, %edx
+        cmpl      $255, %edx
+        je        ..B1.17
+..B1.2:
+        pxor      %xmm7, %xmm7
+        cvtss2sd  (%rdi), %xmm7
+        cmpl      $74, %edx
+        jle       ..B1.15
+..B1.3:
+        movsd     1080+_vmldExpHATab(%rip), %xmm0
+        comisd    %xmm7, %xmm0
+        jb        ..B1.14
+..B1.4:
+        comisd    1096+_vmldExpHATab(%rip), %xmm7
+        jb        ..B1.13
+..B1.5:
+        movsd     1024+_vmldExpHATab(%rip), %xmm1
+        movaps    %xmm7, %xmm6
+        mulsd     %xmm7, %xmm1
+        lea       _vmldExpHATab(%rip), %r10
+        movsd     %xmm1, -16(%rsp)
+        movsd     -16(%rsp), %xmm2
+        movsd     1072+_vmldExpHATab(%rip), %xmm1
+        movsd     1136+_vmldExpHATab(%rip), %xmm0
+        movsd     %xmm0, -24(%rsp)
+        addsd     1032+_vmldExpHATab(%rip), %xmm2
+        movsd     %xmm2, -8(%rsp)
+        movsd     -8(%rsp), %xmm3
+        movl      -8(%rsp), %edi
+        movl      %edi, %ecx
+        andl      $63, %edi
+        subsd     1032+_vmldExpHATab(%rip), %xmm3
+        movsd     %xmm3, -16(%rsp)
+        lea       1(%rdi,%rdi), %r8d
+        movsd     -16(%rsp), %xmm4
+        lea       (%rdi,%rdi), %r9d
+        mulsd     1104+_vmldExpHATab(%rip), %xmm4
+        movsd     -16(%rsp), %xmm5
+        subsd     %xmm4, %xmm6
+        mulsd     1112+_vmldExpHATab(%rip), %xmm5
+        shrl      $6, %ecx
+        subsd     %xmm5, %xmm6
+        comisd    1088+_vmldExpHATab(%rip), %xmm7
+        mulsd     %xmm6, %xmm1
+        movsd     (%r10,%r9,8), %xmm0
+        lea       1023(%rcx), %edx
+        addsd     1064+_vmldExpHATab(%rip), %xmm1
+        mulsd     %xmm6, %xmm1
+        addsd     1056+_vmldExpHATab(%rip), %xmm1
+        mulsd     %xmm6, %xmm1
+        addsd     1048+_vmldExpHATab(%rip), %xmm1
+        mulsd     %xmm6, %xmm1
+        addsd     1040+_vmldExpHATab(%rip), %xmm1
+        mulsd     %xmm6, %xmm1
+        mulsd     %xmm6, %xmm1
+        addsd     %xmm6, %xmm1
+        addsd     (%r10,%r8,8), %xmm1
+        mulsd     %xmm0, %xmm1
+        jb        ..B1.9
+..B1.6:
+        andl      $2047, %edx
+        addsd     %xmm0, %xmm1
+        cmpl      $2046, %edx
+        ja        ..B1.8
+..B1.7:
+        movzwl    1142+_vmldExpHATab(%rip), %ecx
+        shll      $4, %edx
+        andl      $-32753, %ecx
+        orl       %edx, %ecx
+        movw      %cx, -18(%rsp)
+        movsd     -24(%rsp), %xmm0
+        mulsd     %xmm0, %xmm1
+        cvtsd2ss  %xmm1, %xmm1
+        movss     %xmm1, (%rsi)
+        ret
+..B1.8:
+        decl      %edx
+        andl      $2047, %edx
+        movzwl    -18(%rsp), %ecx
+        shll      $4, %edx
+        andl      $-32753, %ecx
+        orl       %edx, %ecx
+        movw      %cx, -18(%rsp)
+        movsd     -24(%rsp), %xmm0
+        mulsd     %xmm0, %xmm1
+        mulsd     1152+_vmldExpHATab(%rip), %xmm1
+        cvtsd2ss  %xmm1, %xmm1
+        movss     %xmm1, (%rsi)
+        ret
+..B1.9:
+        addl      $1083, %ecx
+        andl      $2047, %ecx
+        movl      %ecx, %eax
+        movzwl    -18(%rsp), %edx
+        shll      $4, %eax
+        andl      $-32753, %edx
+        orl       %eax, %edx
+        movw      %dx, -18(%rsp)
+        movsd     -24(%rsp), %xmm3
+        mulsd     %xmm3, %xmm1
+        mulsd     %xmm0, %xmm3
+        movaps    %xmm1, %xmm2
+        addsd     %xmm3, %xmm2
+        cmpl      $50, %ecx
+        ja        ..B1.11
+..B1.10:
+        movsd     1160+_vmldExpHATab(%rip), %xmm0
+        mulsd     %xmm0, %xmm2
+        cvtsd2ss  %xmm2, %xmm2
+        jmp       ..B1.12
+..B1.11:
+        movsd     %xmm2, -72(%rsp)
+        movsd     -72(%rsp), %xmm0
+        subsd     %xmm0, %xmm3
+        movsd     %xmm3, -64(%rsp)
+        movsd     -64(%rsp), %xmm2
+        addsd     %xmm1, %xmm2
+        movsd     %xmm2, -64(%rsp)
+        movsd     -72(%rsp), %xmm1
+        mulsd     1168+_vmldExpHATab(%rip), %xmm1
+        movsd     %xmm1, -56(%rsp)
+        movsd     -72(%rsp), %xmm4
+        movsd     -56(%rsp), %xmm3
+        addsd     %xmm3, %xmm4
+        movsd     %xmm4, -48(%rsp)
+        movsd     -48(%rsp), %xmm6
+        movsd     -56(%rsp), %xmm5
+        subsd     %xmm5, %xmm6
+        movsd     %xmm6, -40(%rsp)
+        movsd     -72(%rsp), %xmm8
+        movsd     -40(%rsp), %xmm7
+        subsd     %xmm7, %xmm8
+        movsd     %xmm8, -32(%rsp)
+        movsd     -64(%rsp), %xmm10
+        movsd     -32(%rsp), %xmm9
+        addsd     %xmm9, %xmm10
+        movsd     %xmm10, -32(%rsp)
+        movsd     -40(%rsp), %xmm11
+        mulsd     1160+_vmldExpHATab(%rip), %xmm11
+        movsd     %xmm11, -40(%rsp)
+        movsd     -32(%rsp), %xmm12
+        mulsd     1160+_vmldExpHATab(%rip), %xmm12
+        movsd     %xmm12, -32(%rsp)
+        movsd     -40(%rsp), %xmm2
+        movsd     -32(%rsp), %xmm13
+        addsd     %xmm13, %xmm2
+        cvtsd2ss  %xmm2, %xmm2
+..B1.12:
+        movss     %xmm2, (%rsi)
+        movl      $4, %eax
+        ret
+..B1.13:
+        movsd     1120+_vmldExpHATab(%rip), %xmm0
+        movl      $4, %eax
+        mulsd     %xmm0, %xmm0
+        cvtsd2ss  %xmm0, %xmm0
+        movss     %xmm0, (%rsi)
+        ret
+..B1.14:
+        movsd     1128+_vmldExpHATab(%rip), %xmm0
+        movl      $3, %eax
+        mulsd     %xmm0, %xmm0
+        cvtsd2ss  %xmm0, %xmm0
+        movss     %xmm0, (%rsi)
+        ret
+..B1.15:
+        movsd     1144+_vmldExpHATab(%rip), %xmm0
+        addsd     %xmm7, %xmm0
+        cvtsd2ss  %xmm0, %xmm0
+        movss     %xmm0, (%rsi)
+..B1.16:
+        ret
+..B1.17:
+        movb      3(%rdi), %dl
+        andb      $-128, %dl
+        cmpb      $-128, %dl
+        je        ..B1.19
+..B1.18:
+        movss     (%rdi), %xmm0
+        mulss     %xmm0, %xmm0
+        movss     %xmm0, (%rsi)
+        ret
+..B1.19:
+        testl     $8388607, (%rdi)
+        jne       ..B1.18
+..B1.20:
+        movsd     1136+_vmldExpHATab(%rip), %xmm0
+        cvtsd2ss  %xmm0, %xmm0
+        movss     %xmm0, (%rsi)
+        ret
+        .align    16,0x90
+	.cfi_endproc
+	.type	__svml_sexp_ha_cout_rare_internal,@function
+	.size	__svml_sexp_ha_cout_rare_internal,.-__svml_sexp_ha_cout_rare_internal
+..LN__svml_sexp_ha_cout_rare_internal.0:
+	.data
+# -- End  __svml_sexp_ha_cout_rare_internal
+	.section .rodata, "a"
+	.align 32
+	.align 32
+_vmldExpHATab:
+	.long	0
+	.long	1072693248
+	.long	0
+	.long	0
+	.long	1048019041
+	.long	1072704666
+	.long	2631457885
+	.long	3161546771
+	.long	3541402996
+	.long	1072716208
+	.long	896005651
+	.long	1015861842
+	.long	410360776
+	.long	1072727877
+	.long	1642514529
+	.long	1012987726
+	.long	1828292879
+	.long	1072739672
+	.long	1568897901
+	.long	1016568486
+	.long	852742562
+	.long	1072751596
+	.long	1882168529
+	.long	1010744893
+	.long	3490863953
+	.long	1072763649
+	.long	707771662
+	.long	3163903570
+	.long	2930322912
+	.long	1072775834
+	.long	3117806614
+	.long	3163670819
+	.long	1014845819
+	.long	1072788152
+	.long	3936719688
+	.long	3162512149
+	.long	3949972341
+	.long	1072800603
+	.long	1058231231
+	.long	1015777676
+	.long	828946858
+	.long	1072813191
+	.long	1044000608
+	.long	1016786167
+	.long	2288159958
+	.long	1072825915
+	.long	1151779725
+	.long	1015705409
+	.long	1853186616
+	.long	1072838778
+	.long	3819481236
+	.long	1016499965
+	.long	1709341917
+	.long	1072851781
+	.long	2552227826
+	.long	1015039787
+	.long	4112506593
+	.long	1072864925
+	.long	1829350193
+	.long	1015216097
+	.long	2799960843
+	.long	1072878213
+	.long	1913391796
+	.long	1015756674
+	.long	171030293
+	.long	1072891646
+	.long	1303423926
+	.long	1015238005
+	.long	2992903935
+	.long	1072905224
+	.long	1574172746
+	.long	1016061241
+	.long	926591435
+	.long	1072918951
+	.long	3427487848
+	.long	3163704045
+	.long	887463927
+	.long	1072932827
+	.long	1049900754
+	.long	3161575912
+	.long	1276261410
+	.long	1072946854
+	.long	2804567149
+	.long	1015390024
+	.long	569847338
+	.long	1072961034
+	.long	1209502043
+	.long	3159926671
+	.long	1617004845
+	.long	1072975368
+	.long	1623370769
+	.long	1011049453
+	.long	3049340112
+	.long	1072989858
+	.long	3667985273
+	.long	1013894369
+	.long	3577096743
+	.long	1073004506
+	.long	3145379760
+	.long	1014403278
+	.long	1990012071
+	.long	1073019314
+	.long	7447438
+	.long	3163526196
+	.long	1453150082
+	.long	1073034283
+	.long	3171891295
+	.long	3162037958
+	.long	917841882
+	.long	1073049415
+	.long	419288974
+	.long	1016280325
+	.long	3712504873
+	.long	1073064711
+	.long	3793507337
+	.long	1016095713
+	.long	363667784
+	.long	1073080175
+	.long	728023093
+	.long	1016345318
+	.long	2956612997
+	.long	1073095806
+	.long	1005538728
+	.long	3163304901
+	.long	2186617381
+	.long	1073111608
+	.long	2018924632
+	.long	3163803357
+	.long	1719614413
+	.long	1073127582
+	.long	3210617384
+	.long	3163796463
+	.long	1013258799
+	.long	1073143730
+	.long	3094194670
+	.long	3160631279
+	.long	3907805044
+	.long	1073160053
+	.long	2119843535
+	.long	3161988964
+	.long	1447192521
+	.long	1073176555
+	.long	508946058
+	.long	3162904882
+	.long	1944781191
+	.long	1073193236
+	.long	3108873501
+	.long	3162190556
+	.long	919555682
+	.long	1073210099
+	.long	2882956373
+	.long	1013312481
+	.long	2571947539
+	.long	1073227145
+	.long	4047189812
+	.long	3163777462
+	.long	2604962541
+	.long	1073244377
+	.long	3631372142
+	.long	3163870288
+	.long	1110089947
+	.long	1073261797
+	.long	3253791412
+	.long	1015920431
+	.long	2568320822
+	.long	1073279406
+	.long	1509121860
+	.long	1014756995
+	.long	2966275557
+	.long	1073297207
+	.long	2339118633
+	.long	3160254904
+	.long	2682146384
+	.long	1073315202
+	.long	586480042
+	.long	3163702083
+	.long	2191782032
+	.long	1073333393
+	.long	730975783
+	.long	1014083580
+	.long	2069751141
+	.long	1073351782
+	.long	576856675
+	.long	3163014404
+	.long	2990417245
+	.long	1073370371
+	.long	3552361237
+	.long	3163667409
+	.long	1434058175
+	.long	1073389163
+	.long	1853053619
+	.long	1015310724
+	.long	2572866477
+	.long	1073408159
+	.long	2462790535
+	.long	1015814775
+	.long	3092190715
+	.long	1073427362
+	.long	1457303226
+	.long	3159737305
+	.long	4076559943
+	.long	1073446774
+	.long	950899508
+	.long	3160987380
+	.long	2420883922
+	.long	1073466398
+	.long	174054861
+	.long	1014300631
+	.long	3716502172
+	.long	1073486235
+	.long	816778419
+	.long	1014197934
+	.long	777507147
+	.long	1073506289
+	.long	3507050924
+	.long	1015341199
+	.long	3706687593
+	.long	1073526560
+	.long	1821514088
+	.long	1013410604
+	.long	1242007932
+	.long	1073547053
+	.long	1073740399
+	.long	3163532637
+	.long	3707479175
+	.long	1073567768
+	.long	2789017511
+	.long	1014276997
+	.long	64696965
+	.long	1073588710
+	.long	3586233004
+	.long	1015962192
+	.long	863738719
+	.long	1073609879
+	.long	129252895
+	.long	3162690849
+	.long	3884662774
+	.long	1073631278
+	.long	1614448851
+	.long	1014281732
+	.long	2728693978
+	.long	1073652911
+	.long	2413007344
+	.long	3163551506
+	.long	3999357479
+	.long	1073674779
+	.long	1101668360
+	.long	1015989180
+	.long	1533953344
+	.long	1073696886
+	.long	835814894
+	.long	1015702697
+	.long	2174652632
+	.long	1073719233
+	.long	1301400989
+	.long	1014466875
+	.long	1697350398
+	.long	1079448903
+	.long	0
+	.long	1127743488
+	.long	0
+	.long	1071644672
+	.long	1431652600
+	.long	1069897045
+	.long	1431670732
+	.long	1067799893
+	.long	984555731
+	.long	1065423122
+	.long	472530941
+	.long	1062650218
+	.long	3758096384
+	.long	1079389762
+	.long	3758096384
+	.long	3226850697
+	.long	2147483648
+	.long	3227123254
+	.long	4277796864
+	.long	1065758274
+	.long	3164486458
+	.long	1025308570
+	.long	1
+	.long	1048576
+	.long	4294967295
+	.long	2146435071
+	.long	0
+	.long	0
+	.long	0
+	.long	1072693248
+	.long	0
+	.long	1073741824
+	.long	0
+	.long	1009778688
+	.long	0
+	.long	1106771968
+	.type	_vmldExpHATab,@object
+	.size	_vmldExpHATab,1176
+	.data
+	.section .note.GNU-stack, ""
+# End
diff --git a/libm/x86_64/expf_data_avx2.S b/libm/x86_64/expf_data_avx2.S
new file mode 100644
index 0000000..f9a98e8
--- /dev/null
+++ b/libm/x86_64/expf_data_avx2.S
@@ -0,0 +1,958 @@
+/*
+Copyright (C) 2019 The Android Open Source Project
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions
+are met:
+ * Redistributions of source code must retain the above copyright
+   notice, this list of conditions and the following disclaimer.
+ * Redistributions in binary form must reproduce the above copyright
+   notice, this list of conditions and the following disclaimer in
+   the documentation and/or other materials provided with the
+   distribution.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT
+OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+SUCH DAMAGE.
+*/
+
+# -- Machine type EM64t
+	.section .rodata, "a"
+	.align 64
+	.align 64
+	.hidden __svml_sexp_ha_data_internal_avx512
+	.globl __svml_sexp_ha_data_internal_avx512
+__svml_sexp_ha_data_internal_avx512:
+	.long	0
+	.long	3007986186
+	.long	860277610
+	.long	3010384254
+	.long	2991457809
+	.long	3008462297
+	.long	860562562
+	.long	3004532446
+	.long	856238081
+	.long	3001480295
+	.long	857441778
+	.long	815380209
+	.long	3003456168
+	.long	3001196762
+	.long	2986372182
+	.long	3006683458
+	.long	848495278
+	.long	851809756
+	.long	3003311522
+	.long	2995654817
+	.long	833868005
+	.long	3004843819
+	.long	835836658
+	.long	3003498340
+	.long	2994528642
+	.long	3002229827
+	.long	2981408986
+	.long	2983889551
+	.long	2983366846
+	.long	3000350873
+	.long	833659207
+	.long	2987748092
+	.long	1065353216
+	.long	1065536903
+	.long	1065724611
+	.long	1065916431
+	.long	1066112450
+	.long	1066312762
+	.long	1066517459
+	.long	1066726640
+	.long	1066940400
+	.long	1067158842
+	.long	1067382066
+	.long	1067610179
+	.long	1067843287
+	.long	1068081499
+	.long	1068324927
+	.long	1068573686
+	.long	1068827891
+	.long	1069087663
+	.long	1069353124
+	.long	1069624397
+	.long	1069901610
+	.long	1070184894
+	.long	1070474380
+	.long	1070770206
+	.long	1071072509
+	.long	1071381432
+	.long	1071697119
+	.long	1072019719
+	.long	1072349383
+	.long	1072686266
+	.long	1073030525
+	.long	1073382323
+	.long	1069066811
+	.long	1069066811
+	.long	1069066811
+	.long	1069066811
+	.long	1069066811
+	.long	1069066811
+	.long	1069066811
+	.long	1069066811
+	.long	1069066811
+	.long	1069066811
+	.long	1069066811
+	.long	1069066811
+	.long	1069066811
+	.long	1069066811
+	.long	1069066811
+	.long	1069066811
+	.long	1220542464
+	.long	1220542464
+	.long	1220542464
+	.long	1220542464
+	.long	1220542464
+	.long	1220542464
+	.long	1220542464
+	.long	1220542464
+	.long	1220542464
+	.long	1220542464
+	.long	1220542464
+	.long	1220542464
+	.long	1220542464
+	.long	1220542464
+	.long	1220542464
+	.long	1220542464
+	.long	1060205080
+	.long	1060205080
+	.long	1060205080
+	.long	1060205080
+	.long	1060205080
+	.long	1060205080
+	.long	1060205080
+	.long	1060205080
+	.long	1060205080
+	.long	1060205080
+	.long	1060205080
+	.long	1060205080
+	.long	1060205080
+	.long	1060205080
+	.long	1060205080
+	.long	1060205080
+	.long	2969756424
+	.long	2969756424
+	.long	2969756424
+	.long	2969756424
+	.long	2969756424
+	.long	2969756424
+	.long	2969756424
+	.long	2969756424
+	.long	2969756424
+	.long	2969756424
+	.long	2969756424
+	.long	2969756424
+	.long	2969756424
+	.long	2969756424
+	.long	2969756424
+	.long	2969756424
+	.long	3221225471
+	.long	3221225471
+	.long	3221225471
+	.long	3221225471
+	.long	3221225471
+	.long	3221225471
+	.long	3221225471
+	.long	3221225471
+	.long	3221225471
+	.long	3221225471
+	.long	3221225471
+	.long	3221225471
+	.long	3221225471
+	.long	3221225471
+	.long	3221225471
+	.long	3221225471
+	.long	1042983923
+	.long	1042983923
+	.long	1042983923
+	.long	1042983923
+	.long	1042983923
+	.long	1042983923
+	.long	1042983923
+	.long	1042983923
+	.long	1042983923
+	.long	1042983923
+	.long	1042983923
+	.long	1042983923
+	.long	1042983923
+	.long	1042983923
+	.long	1042983923
+	.long	1042983923
+	.long	1056964854
+	.long	1056964854
+	.long	1056964854
+	.long	1056964854
+	.long	1056964854
+	.long	1056964854
+	.long	1056964854
+	.long	1056964854
+	.long	1056964854
+	.long	1056964854
+	.long	1056964854
+	.long	1056964854
+	.long	1056964854
+	.long	1056964854
+	.long	1056964854
+	.long	1056964854
+	.long	2147483647
+	.long	2147483647
+	.long	2147483647
+	.long	2147483647
+	.long	2147483647
+	.long	2147483647
+	.long	2147483647
+	.long	2147483647
+	.long	2147483647
+	.long	2147483647
+	.long	2147483647
+	.long	2147483647
+	.long	2147483647
+	.long	2147483647
+	.long	2147483647
+	.long	2147483647
+	.long	1118743631
+	.long	1118743631
+	.long	1118743631
+	.long	1118743631
+	.long	1118743631
+	.long	1118743631
+	.long	1118743631
+	.long	1118743631
+	.long	1118743631
+	.long	1118743631
+	.long	1118743631
+	.long	1118743631
+	.long	1118743631
+	.long	1118743631
+	.long	1118743631
+	.long	1118743631
+	.long	796917760
+	.long	796917760
+	.long	796917760
+	.long	796917760
+	.long	796917760
+	.long	796917760
+	.long	796917760
+	.long	796917760
+	.long	796917760
+	.long	796917760
+	.long	796917760
+	.long	796917760
+	.long	796917760
+	.long	796917760
+	.long	796917760
+	.long	796917760
+	.long	124
+	.long	124
+	.long	124
+	.long	124
+	.long	124
+	.long	124
+	.long	124
+	.long	124
+	.long	124
+	.long	124
+	.long	124
+	.long	124
+	.long	124
+	.long	124
+	.long	124
+	.long	124
+	.type	__svml_sexp_ha_data_internal_avx512,@object
+	.size	__svml_sexp_ha_data_internal_avx512,960
+	.align 64
+	.hidden __svml_sexp_ha_data_internal
+	.globl __svml_sexp_ha_data_internal
+__svml_sexp_ha_data_internal:
+	.long	1085844027
+	.long	1085844027
+	.long	1085844027
+	.long	1085844027
+	.long	1085844027
+	.long	1085844027
+	.long	1085844027
+	.long	1085844027
+	.long	1085844027
+	.long	1085844027
+	.long	1085844027
+	.long	1085844027
+	.long	1085844027
+	.long	1085844027
+	.long	1085844027
+	.long	1085844027
+	.long	1262486012
+	.long	1262486012
+	.long	1262486012
+	.long	1262486012
+	.long	1262486012
+	.long	1262486012
+	.long	1262486012
+	.long	1262486012
+	.long	1262486012
+	.long	1262486012
+	.long	1262486012
+	.long	1262486012
+	.long	1262486012
+	.long	1262486012
+	.long	1262486012
+	.long	1262486012
+	.long	2147483647
+	.long	2147483647
+	.long	2147483647
+	.long	2147483647
+	.long	2147483647
+	.long	2147483647
+	.long	2147483647
+	.long	2147483647
+	.long	2147483647
+	.long	2147483647
+	.long	2147483647
+	.long	2147483647
+	.long	2147483647
+	.long	2147483647
+	.long	2147483647
+	.long	2147483647
+	.long	1118743631
+	.long	1118743631
+	.long	1118743631
+	.long	1118743631
+	.long	1118743631
+	.long	1118743631
+	.long	1118743631
+	.long	1118743631
+	.long	1118743631
+	.long	1118743631
+	.long	1118743631
+	.long	1118743631
+	.long	1118743631
+	.long	1118743631
+	.long	1118743631
+	.long	1118743631
+	.long	1043427328
+	.long	1043427328
+	.long	1043427328
+	.long	1043427328
+	.long	1043427328
+	.long	1043427328
+	.long	1043427328
+	.long	1043427328
+	.long	1043427328
+	.long	1043427328
+	.long	1043427328
+	.long	1043427328
+	.long	1043427328
+	.long	1043427328
+	.long	1043427328
+	.long	1043427328
+	.long	923139572
+	.long	923139572
+	.long	923139572
+	.long	923139572
+	.long	923139572
+	.long	923139572
+	.long	923139572
+	.long	923139572
+	.long	923139572
+	.long	923139572
+	.long	923139572
+	.long	923139572
+	.long	923139572
+	.long	923139572
+	.long	923139572
+	.long	923139572
+	.long	0
+	.long	856238081
+	.long	848495278
+	.long	2994528642
+	.long	0
+	.long	856238081
+	.long	848495278
+	.long	2994528642
+	.long	0
+	.long	856238081
+	.long	848495278
+	.long	2994528642
+	.long	0
+	.long	856238081
+	.long	848495278
+	.long	2994528642
+	.long	1065353216
+	.long	1066940400
+	.long	1068827891
+	.long	1071072509
+	.long	1065353216
+	.long	1066940400
+	.long	1068827891
+	.long	1071072509
+	.long	1065353216
+	.long	1066940400
+	.long	1068827891
+	.long	1071072509
+	.long	1065353216
+	.long	1066940400
+	.long	1068827891
+	.long	1071072509
+	.long	1026210578
+	.long	1026210578
+	.long	1026210578
+	.long	1026210578
+	.long	1026210578
+	.long	1026210578
+	.long	1026210578
+	.long	1026210578
+	.long	1026210578
+	.long	1026210578
+	.long	1026210578
+	.long	1026210578
+	.long	1026210578
+	.long	1026210578
+	.long	1026210578
+	.long	1026210578
+	.long	1042987794
+	.long	1042987794
+	.long	1042987794
+	.long	1042987794
+	.long	1042987794
+	.long	1042987794
+	.long	1042987794
+	.long	1042987794
+	.long	1042987794
+	.long	1042987794
+	.long	1042987794
+	.long	1042987794
+	.long	1042987794
+	.long	1042987794
+	.long	1042987794
+	.long	1042987794
+	.long	1056964607
+	.long	1056964607
+	.long	1056964607
+	.long	1056964607
+	.long	1056964607
+	.long	1056964607
+	.long	1056964607
+	.long	1056964607
+	.long	1056964607
+	.long	1056964607
+	.long	1056964607
+	.long	1056964607
+	.long	1056964607
+	.long	1056964607
+	.long	1056964607
+	.long	1056964607
+	.long	2139095040
+	.long	2139095040
+	.long	2139095040
+	.long	2139095040
+	.long	2139095040
+	.long	2139095040
+	.long	2139095040
+	.long	2139095040
+	.long	2139095040
+	.long	2139095040
+	.long	2139095040
+	.long	2139095040
+	.long	2139095040
+	.long	2139095040
+	.long	2139095040
+	.long	2139095040
+	.long	1065353216
+	.long	0
+	.long	1065398765
+	.long	857715507
+	.long	1065444562
+	.long	2982428624
+	.long	1065490607
+	.long	856090616
+	.long	1065536903
+	.long	3008276393
+	.long	1065583449
+	.long	857595758
+	.long	1065630249
+	.long	3002158215
+	.long	1065677302
+	.long	858550072
+	.long	1065724611
+	.long	860854417
+	.long	1065772177
+	.long	859251839
+	.long	1065820002
+	.long	3008545510
+	.long	1065868085
+	.long	864021810
+	.long	1065916431
+	.long	3011435083
+	.long	1065965038
+	.long	819487640
+	.long	1066013909
+	.long	862622430
+	.long	1066063046
+	.long	856316133
+	.long	1066112450
+	.long	2992679841
+	.long	1066162122
+	.long	3001970243
+	.long	1066212063
+	.long	859241655
+	.long	1066262276
+	.long	847285146
+	.long	1066312762
+	.long	3010032741
+	.long	1066363521
+	.long	843003463
+	.long	1066414556
+	.long	856041677
+	.long	1066465869
+	.long	3009946167
+	.long	1066517459
+	.long	862410276
+	.long	1066569330
+	.long	863709796
+	.long	1066621483
+	.long	851201948
+	.long	1066673919
+	.long	838513088
+	.long	1066726640
+	.long	3006136850
+	.long	1066779647
+	.long	3010613863
+	.long	1066832941
+	.long	853311585
+	.long	1066886525
+	.long	856941301
+	.long	1066940400
+	.long	857938801
+	.long	1066994568
+	.long	2993474036
+	.long	1067049030
+	.long	3009003152
+	.long	1067103787
+	.long	2967126046
+	.long	1067158842
+	.long	3003929955
+	.long	1067214195
+	.long	860904838
+	.long	1067269850
+	.long	3006375425
+	.long	1067325806
+	.long	843377209
+	.long	1067382066
+	.long	859906882
+	.long	1067438632
+	.long	848662531
+	.long	1067495505
+	.long	838571797
+	.long	1067552687
+	.long	3000363757
+	.long	1067610179
+	.long	818090586
+	.long	1067667983
+	.long	860999685
+	.long	1067726102
+	.long	3005475891
+	.long	1067784536
+	.long	3010626242
+	.long	1067843287
+	.long	3006045534
+	.long	1067902357
+	.long	2998944862
+	.long	1067961748
+	.long	3004575030
+	.long	1068021461
+	.long	842258832
+	.long	1068081499
+	.long	3004574472
+	.long	1068141862
+	.long	858118433
+	.long	1068202554
+	.long	3004476802
+	.long	1068263575
+	.long	3006694272
+	.long	1068324927
+	.long	2989353718
+	.long	1068386612
+	.long	858100843
+	.long	1068448633
+	.long	3005558251
+	.long	1068510990
+	.long	3000050815
+	.long	1068573686
+	.long	3011271336
+	.long	1068636722
+	.long	3006477262
+	.long	1068700100
+	.long	840255625
+	.long	1068763823
+	.long	3007001961
+	.long	1068827891
+	.long	852486010
+	.long	1068892308
+	.long	3006218836
+	.long	1068957074
+	.long	2993076596
+	.long	1069022192
+	.long	3000356208
+	.long	1069087663
+	.long	856606199
+	.long	1069153490
+	.long	856315927
+	.long	1069219675
+	.long	3004946819
+	.long	1069286218
+	.long	863888852
+	.long	1069353124
+	.long	3007401960
+	.long	1069420392
+	.long	832069785
+	.long	1069488026
+	.long	3004369690
+	.long	1069556027
+	.long	3007061599
+	.long	1069624397
+	.long	3000395326
+	.long	1069693138
+	.long	851736822
+	.long	1069762253
+	.long	2996268450
+	.long	1069831743
+	.long	2999890733
+	.long	1069901610
+	.long	839559223
+	.long	1069971857
+	.long	3004476343
+	.long	1070042485
+	.long	3000848665
+	.long	1070113496
+	.long	859381756
+	.long	1070184894
+	.long	3010667426
+	.long	1070256678
+	.long	859604257
+	.long	1070328853
+	.long	2953234114
+	.long	1070401420
+	.long	3010682756
+	.long	1070474380
+	.long	841546788
+	.long	1070547737
+	.long	2999163804
+	.long	1070621492
+	.long	2996061011
+	.long	1070695647
+	.long	860136147
+	.long	1070770206
+	.long	3009158570
+	.long	1070845169
+	.long	3008165066
+	.long	1070920539
+	.long	3005767716
+	.long	1070996318
+	.long	2955948569
+	.long	1071072509
+	.long	3000280458
+	.long	1071149113
+	.long	857267046
+	.long	1071226134
+	.long	3002473793
+	.long	1071303572
+	.long	861820308
+	.long	1071381432
+	.long	3008383516
+	.long	1071459714
+	.long	3010850715
+	.long	1071538421
+	.long	3011199971
+	.long	1071617555
+	.long	2995067967
+	.long	1071697119
+	.long	2988354393
+	.long	1071777115
+	.long	2889750092
+	.long	1071857545
+	.long	862289173
+	.long	1071938413
+	.long	3007287685
+	.long	1072019719
+	.long	2990808038
+	.long	1072101467
+	.long	3005883328
+	.long	1072183658
+	.long	857766040
+	.long	1072266296
+	.long	855693471
+	.long	1072349383
+	.long	2990610142
+	.long	1072432921
+	.long	3004393415
+	.long	1072516912
+	.long	846646433
+	.long	1072601360
+	.long	3008357562
+	.long	1072686266
+	.long	3007858250
+	.long	1072771633
+	.long	3006309869
+	.long	1072857463
+	.long	854681599
+	.long	1072943760
+	.long	2995197552
+	.long	1073030525
+	.long	841557046
+	.long	1073117761
+	.long	861464191
+	.long	1073205471
+	.long	862328226
+	.long	1073293658
+	.long	3003728983
+	.long	1073382323
+	.long	2995926872
+	.long	1073471469
+	.long	861522913
+	.long	1073561100
+	.long	835668076
+	.long	1073651217
+	.long	859371295
+	.long	1127787067
+	.long	1127787067
+	.long	1127787067
+	.long	1127787067
+	.long	1127787067
+	.long	1127787067
+	.long	1127787067
+	.long	1127787067
+	.long	1127787067
+	.long	1127787067
+	.long	1127787067
+	.long	1127787067
+	.long	1127787067
+	.long	1127787067
+	.long	1127787067
+	.long	1127787067
+	.long	1262485504
+	.long	1262485504
+	.long	1262485504
+	.long	1262485504
+	.long	1262485504
+	.long	1262485504
+	.long	1262485504
+	.long	1262485504
+	.long	1262485504
+	.long	1262485504
+	.long	1262485504
+	.long	1262485504
+	.long	1262485504
+	.long	1262485504
+	.long	1262485504
+	.long	1262485504
+	.long	1001488384
+	.long	1001488384
+	.long	1001488384
+	.long	1001488384
+	.long	1001488384
+	.long	1001488384
+	.long	1001488384
+	.long	1001488384
+	.long	1001488384
+	.long	1001488384
+	.long	1001488384
+	.long	1001488384
+	.long	1001488384
+	.long	1001488384
+	.long	1001488384
+	.long	1001488384
+	.long	3051257987
+	.long	3051257987
+	.long	3051257987
+	.long	3051257987
+	.long	3051257987
+	.long	3051257987
+	.long	3051257987
+	.long	3051257987
+	.long	3051257987
+	.long	3051257987
+	.long	3051257987
+	.long	3051257987
+	.long	3051257987
+	.long	3051257987
+	.long	3051257987
+	.long	3051257987
+	.long	16256
+	.long	16256
+	.long	16256
+	.long	16256
+	.long	16256
+	.long	16256
+	.long	16256
+	.long	16256
+	.long	16256
+	.long	16256
+	.long	16256
+	.long	16256
+	.long	16256
+	.long	16256
+	.long	16256
+	.long	16256
+	.long	2139095040
+	.long	2139095040
+	.long	2139095040
+	.long	2139095040
+	.long	2139095040
+	.long	2139095040
+	.long	2139095040
+	.long	2139095040
+	.long	2139095040
+	.long	2139095040
+	.long	2139095040
+	.long	2139095040
+	.long	2139095040
+	.long	2139095040
+	.long	2139095040
+	.long	2139095040
+	.long	127
+	.long	127
+	.long	127
+	.long	127
+	.long	127
+	.long	127
+	.long	127
+	.long	127
+	.long	127
+	.long	127
+	.long	127
+	.long	127
+	.long	127
+	.long	127
+	.long	127
+	.long	127
+	.long	4294967232
+	.long	4294967232
+	.long	4294967232
+	.long	4294967232
+	.long	4294967232
+	.long	4294967232
+	.long	4294967232
+	.long	4294967232
+	.long	4294967232
+	.long	4294967232
+	.long	4294967232
+	.long	4294967232
+	.long	4294967232
+	.long	4294967232
+	.long	4294967232
+	.long	4294967232
+	.long	864026624
+	.long	864026624
+	.long	864026624
+	.long	864026624
+	.long	864026624
+	.long	864026624
+	.long	864026624
+	.long	864026624
+	.long	864026624
+	.long	864026624
+	.long	864026624
+	.long	864026624
+	.long	864026624
+	.long	864026624
+	.long	864026624
+	.long	864026624
+	.long	1056964608
+	.long	1056964608
+	.long	1056964608
+	.long	1056964608
+	.long	1056964608
+	.long	1056964608
+	.long	1056964608
+	.long	1056964608
+	.long	1056964608
+	.long	1056964608
+	.long	1056964608
+	.long	1056964608
+	.long	1056964608
+	.long	1056964608
+	.long	1056964608
+	.long	1056964608
+	.long	2147483647
+	.long	2147483647
+	.long	2147483647
+	.long	2147483647
+	.long	2147483647
+	.long	2147483647
+	.long	2147483647
+	.long	2147483647
+	.long	2147483647
+	.long	2147483647
+	.long	2147483647
+	.long	2147483647
+	.long	2147483647
+	.long	2147483647
+	.long	2147483647
+	.long	2147483647
+	.long	1118743631
+	.long	1118743631
+	.long	1118743631
+	.long	1118743631
+	.long	1118743631
+	.long	1118743631
+	.long	1118743631
+	.long	1118743631
+	.long	1118743631
+	.long	1118743631
+	.long	1118743631
+	.long	1118743631
+	.long	1118743631
+	.long	1118743631
+	.long	1118743631
+	.long	1118743631
+	.long	1118925335
+	.long	1118925335
+	.long	1118925335
+	.long	1118925335
+	.long	1118925335
+	.long	1118925335
+	.long	1118925335
+	.long	1118925335
+	.long	1118925335
+	.long	1118925335
+	.long	1118925335
+	.long	1118925335
+	.long	1118925335
+	.long	1118925335
+	.long	1118925335
+	.long	1118925335
+	.long	3268407732
+	.long	3268407732
+	.long	3268407732
+	.long	3268407732
+	.long	3268407732
+	.long	3268407732
+	.long	3268407732
+	.long	3268407732
+	.long	3268407732
+	.long	3268407732
+	.long	3268407732
+	.long	3268407732
+	.long	3268407732
+	.long	3268407732
+	.long	3268407732
+	.long	3268407732
+	.type	__svml_sexp_ha_data_internal,@object
+	.size	__svml_sexp_ha_data_internal,2688
+	.data
+	.section .note.GNU-stack, ""
+# End
-- 
2.7.4

